source/common/http/http1/parser.h

```diff
@@ -0,0 +1,183 @@
+   * @param data supplies the start address.
+   * @param length supplies the length
+   */
+  virtual void bufferBody(const char* data, size_t length) PURE;
+
+  /**
+   * Called when the HTTP message has completed parsing.
+   * @return An error status or a ParserStatus.
+   */
+  virtual StatusOr<ParserStatus> onMessageComplete() PURE;
+
+  /**
+   * Called when accepting a chunk header.
+   */
+  virtual void onChunkHeader(bool) PURE;
+
+  virtual int setAndCheckCallbackStatus(Status&& status) PURE;
+  virtual int setAndCheckCallbackStatusOr(Envoy::StatusOr<ParserStatus>&& statusor) PURE;
+};
+
+class Parser {
+public:
+  // Struct containing the return value from parser execution.
+  struct RcVal {
+    // Number of parsed bytes.
+    size_t nread;
+    // Integer error from parser indicating return code.
+    int rc;
+  };
+  virtual ~Parser() = default;
+
+  // Executes the parser.
+  // @return an RcVal containing the number of parsed bytes and return code.
+  virtual RcVal execute(const char* slice, int len) PURE;
+
+  // Unpauses the parser.
+  virtual void resume() PURE;
+
+  // Pauses the parser and returns a status indicating pause.
+  virtual ParserStatus pause() PURE;
+
+  // Returns a parser status representing the errno value from the parser.
+  virtual ParserStatus getStatus() PURE;
+
+  // Returns an integer representing the status code stored in the parser structure. For responses
+  // only.
+  virtual uint16_t statusCode() const PURE;
+
+  // Returns an integer representing the HTTP major version.
+  virtual int httpMajor() const PURE;
+
+  // Returns an integer representing the HTTP minor version.
+  virtual int httpMinor() const PURE;
+
+  // Returns the number of bytes in the body. -1 if no Content-Length header
+  virtual uint64_t contentLength() const PURE;
+
+  // Returns parser flags (e.g. chunked).
+  virtual int flags() const PURE;
+
+  // Returns an integer representing the method. For requests only.
+  virtual uint16_t method() const PURE;
+
+  // Returns a textual representation of the method. For requests only.
+  virtual const char* methodName() const PURE;
+
+  // Returns a textual representation of the latest return error.
+  virtual const char* errnoName() PURE;
+
+  // Returns a textual representation of the given return code.
+  virtual const char* errnoName(int rc) const PURE;
+
+  // Returns whether the Transfer-Encoding header is present.
+  virtual int usesTransferEncoding() const PURE;
+
+  // Returns whether the Content-Length header is present.
+  virtual bool seenContentLength() const PURE;
+
+  // Tells the parser that the Content-Length header is present.
+  virtual void setSeenContentLength(bool val) PURE;
+
+  // Converts a ParserStatus code to the parsers' integer return code value.
+  virtual int statusToInt(const ParserStatus code) const PURE;
+
+  // The value of the chunked flag.
+  virtual int flagsChunked() const PURE;
+};
+
+enum class Method {
+  Head = 2,
+  Connect = 5,
```

Comment:
Again should these use the http parser constants (cast if necessary)?
Also are they guaranteed to be consistent for parser next and parser legacy?  should they be functions as well?

Reply:
They (these Methods the codec actually cares about) are consistent between the new and old parser. We could actually eliminate these if the codec compared with method strings... that might be better than making the parser interface depend on the impl's constants